core/src/main/scala/org/apache/spark/h2o/H2OContext.scala

```diff
@@ -106,77 +105,42 @@ abstract class H2OContext private(private val conf: H2OConf) extends Logging wit
-   * otherwise it creates new H2O cluster living in Spark
-   */
-  protected def init(): H2OContext = {
-    logInfo("Sparkling Water version: " + BuildInfo.SWVersion)
-    logInfo("Spark version: " + sparkContext.version)
-    logInfo("Integrated H2O version: " + BuildInfo.H2OVersion)
-    logInfo("The following Spark configuration is used: \n    " + conf.getAll.mkString("\n    "))
-    if (!isRunningOnCorrectSpark(sparkContext)) {
-      throw new WrongSparkVersion(s"You are trying to use Sparkling Water built for Spark ${BuildInfo.buildSparkMajorVersion}," +
-        s" but your $$SPARK_HOME(=${sparkContext.getSparkHome().getOrElse("SPARK_HOME is not defined!")}) property" +
-        s" points to Spark of version ${sparkContext.version}. Please ensure correct Spark is provided and" +
-        s" re-run Sparkling Water.")
-    }
-
-    // Init the H2O Context in a way provided by used backend and return the list of H2O nodes in case of external
-    // backend or list of spark executors on which H2O runs in case of internal backend
-    initBackend()
-    setTimeZone()
-    // The lowest priority used by Spark is 25 (removing temp dirs). We need to perform cleaning up of H2O
-    // resources before Spark does as we run as embedded application inside the Spark
-    shutdownHookRef = ShutdownHookManager.addShutdownHook(10) { () =>
-      logWarning("Spark shutdown hook called, stopping H2OContext!")
-      stop(stopSparkContext = false, stopJvm = false, inShutdownHook = true)
-    }
-    localClientIp = getH2OEndpointIp()
-    localClientPort = getH2OEndpointPort()
-
-    if (sparkContext.ui.isDefined) {
-      SparkSpecificUtils.addSparklingWaterTab(sparkContext)
-    }
-
-    logInfo(s"Sparkling Water ${BuildInfo.SWVersion} started, status of context: $this ")
-    updateUIAfterStart() // updates the spark UI
-    backendHeartbeatThread.start() // start backend heartbeats
-    this
+  H2OContext.logStartingInfo(conf)
+  H2OContext.verifySparkVersion()
+  private val nodes = backend.init(conf)
+  setTimeZone()
+  // The lowest priority used by Spark is 25 (removing temp dirs). We need to perform cleaning up of H2O
+  // resources before Spark does as we run as embedded application inside the Spark
+  private val shutdownHookRef = ShutdownHookManager.addShutdownHook(10) { () =>
+    logWarning("Spark shutdown hook called, stopping H2OContext!")
+    stop(stopSparkContext = false, stopJvm = false, inShutdownHook = true)
   }
+  private val leaderNode = RestApiUtils.getLeaderNode(conf)
+  if (sparkContext.ui.isDefined) {
+    SparkSpecificUtils.addSparklingWaterTab(sparkContext)
+  }
+  logInfo(s"Sparkling Water ${BuildInfo.SWVersion} started, status of context: $this ")
+  updateUIAfterStart() // updates the spark UI
+  backendHeartbeatThread.start() // start backend heartbeats
+  private val (flowIp, flowPort) = {
+    val uri = ProxyStarter.startFlowProxy(conf)
```

Comment:
Always start flow proxy so we have consistent behaviour in both rest api mode and non rest api.